### PR TITLE
safeloader: Switch to "recursive" discovery by default [v3]

### PR DESCRIPTION
**Pull request by ldoktor**
This PR shuffles the loader code quite a lot to allow recursive discovery by default. It keeps the current implementation, therefor on "disable" and "enable" it simply proceeds to discover the class, only remembers whether it seen such tag (as it means it is "avocado.Test" no matter whether it really is).

This code is not perfect, it lacks some extracted attributes to avoid duplicity, but I think it is a standalone unit that could be applied and those extractions and later even optimizations could be done later.

I have not updated documentation yet, please let me know whether this is acceptable and then I'll include documentation fixes.

v1: https://github.com/avocado-framework/avocado/pull/2687
v2: https://github.com/avocado-framework/avocado/pull/2768

Changes:

```yaml
v2: rebased (which resulted in all code being in safeloader rather than loader)
v3: rebased (several commits were already merged)
```

### CODE REVIEW

**Comment by ldoktor** [comment]
Now this is an interesting issue: https://travis-ci.org/avocado-framework/avocado/jobs/427589494#L5146 the order of discovered tests differs. I thought it's suppose to be ordered, but it looks like on some versions of python it can differ... @clebergnu would you please review this and let me know whether it's a bug, or whether I should ignore the order of discovered tests...

**Comment by clebergnu** [comment]
@ldoktor line 5146 contains:

```
selftests/checkall: line 82: wait: pid 57075 is not a child of this shell
```

(among many other similar lines).

And the job finished successfully, which should have been caught by a test failure.  Maybe this is an intermittent behavior?

**Comment by ldoktor** [comment]
that issue is not related to this PR, I saw it multiple times elsewhere (never on my machine). Anyway as mentioned several times there is the ordering issue and missing documentation. Would someone please review the PR and let me know whether I could proceed in this direction, or whether it's unacceptable at all?

**Comment by clebergnu** [comment]
@ldoktor that's exactly what I said.  My point is that you described pointed to line "5146", which clearly did not contain (any more?) the failure you were referring to.

I'm already reviewing this PR, and proposed one of the commits as a hotfix: 2868

**Comment by clebergnu** [comment]
@ldoktor now I've sent #2870, which fixes the bugs you've uncovered.  Now, we have two commits only to go over... I'm already reviewing them.  We're getting there!

File: avocado/core/safeloader.py [hunk 1]
```diff
@@ -30,38 +31,57 @@ class AvocadoModule(object):
...
-
-                for name in statement.names:
-                    if name.name == 'Test':
-                        if name.asname is not None:
-                            self.test_import = name.asname
-                        else:
-                            self.test_import = name.name
-                        break
+            if isinstance(statement, ast.ImportFrom):
+                self.add_imported_object(statement)
```
Comment (clebergnu) [inline avocado/core/safeloader.py]: It looks like `add_import_object` loops the statements the same way as its done two lines below.  It seems like a good candidate for breaking into a reusable function.

File: avocado/core/safeloader.py [hunk 2]
```diff
@@ -250,6 +297,18 @@ def _examine_class(path, class_name):
 
         docstring = ast.get_docstring(klass)
         cl_tags = get_docstring_directives_tags(docstring)
+
+        # Only detect 'avocado.Test' if not yet decided
+        if is_avocado is False:
+            if check_docstring_directive(docstring, 'disable'):
+                is_avocado = True
```
Comment (clebergnu) [inline avocado/core/safeloader.py]: Is this correct? In a quick glance it looks like it should be `False` for a class that 
 contains an explicitl `disable`.

Also, every check_docstring_directive() does a lot of the same work.  Maybe it would be better to get all the directives once:

```python
directives = get_docstring_directives(docstring)
if 'disable' in directives:
   is_avocado = ...
```

File: avocado/core/safeloader.py [hunk 3]
```diff
@@ -264,59 +323,58 @@ def _examine_class(path, class_name):
...
                 continue
             parent_class = parent.id
-            _info, _disabled = _examine_class(path, parent_class)
+            _info, _disabled, _avocado = _examine_class(path, parent_class,
+                                                        is_avocado)
             if _info:
                 parents.remove(parent)
                 info.extend(_info)
                 disabled.update(_disabled)
+            if _avocado is not is_avocado:
```
Comment (clebergnu) [inline avocado/core/safeloader.py]: I know what you're up to here, but this amount of similarly named variables makes readability a bit tough.  Suggestion is to increase the verbosity of names if necessary to increase readability and maintainability.

File: avocado/core/safeloader.py [hunk 4]
```diff
@@ -30,38 +31,57 @@ class AvocadoModule(object):
...
+        #            Basicall it's $path/$module/$variable, but depending
+        #            on import could be also $path/$module.
+        self.imported_objects = {}
         if os.path.isdir(path):
-            self.path = os.path.join(path, "__init__.py")
-        else:
-            self.path = path
-        with open(self.path) as source_file:
-            self.mod = ast.parse(source_file.read(), self.path)
+            path = os.path.join(path, "__init__.py")
```
Comment (clebergnu) [inline avocado/core/safeloader.py]: This simplification of the path assignment can be extract into its own commit.

File: avocado/core/safeloader.py (hunk 2 above)
Comment (ldoktor) [reply avocado/core/safeloader.py]: The `is_avocado` means that the class is discovered to be avocado entity. This is very important for recursive discovery (and was buggy in recursive discovery...)

Comment (ldoktor) [reply avocado/core/safeloader.py]: ... meaning it is avocado entity, it's just not a test.

Comment (ldoktor) [reply avocado/core/safeloader.py]: Oh, as for the optimization, sure, could that wait after we get recursive discovery by default?

File: avocado/core/safeloader.py (hunk 3 above)
Comment (ldoktor) [reply avocado/core/safeloader.py]: Well the `_` variables are looping/temporary variables. It does not look confusing to me. But if you have a better suggestion, I can refactor that.

**Comment by ldoktor** [comment]
@clebergnu there were a few comments, I tried answering all of them but not sure whether I missed some. Would you please clearly mention the necessary steps here (alternatively let's just follow up with the responded ones and I can then send v4 where we can clear the remaining bits as this is getting a bit dirty).

File: avocado/core/safeloader.py (hunk 4 above)
Comment (ldoktor) [reply avocado/core/safeloader.py]: It could, similarly to number of other little refactors or improvements but do you think it's that important? I mean do you expect partial reverts or would it speedup the review process? I don't think so, anyway if you do I can extract it. It's a matter of a few clicks, commit message, another ci check and another review...

**Comment by ldoktor** [comment]
@clebergnu, would you please at least follow-up the answered comments?

File: avocado/core/safeloader.py (hunk 1 above)
Comment (clebergnu) [reply avocado/core/safeloader.py]: You're right, it may add complexity to the code.

File: avocado/core/safeloader.py [hunk 5]
```diff
@@ -227,7 +247,32 @@ def get_methods_info(statement_body, class_tags):
     return methods_info
 
 
-def _examine_class(path, class_name):
+def _is_class_avocado_test(module, klass):
```
Comment (clebergnu) [inline avocado/core/safeloader.py]: Different than, say, `module_imported_as()`, which takes a `module` parameter that is the result for `ast.parse()`, this `_is_class_avocado_test()` utility function always operates on the basis of a module parameter that is an `AvocadoModule` instance.

So, it makes sense to have this as an utility *method*, of the `AvocadoModule` class instead.  Doing so also eliminates the confusion between what the `module` parameter here means.

Comment (clebergnu) [reply avocado/core/safeloader.py]: What I mean is best shown here:

https://github.com/avocado-framework/avocado/pull/3008/commits/9b6646c47db2b8e50e18bf4d60155cfea99b0eef

File: avocado/core/safeloader.py (hunk 4 above)
Comment (clebergnu) [reply avocado/core/safeloader.py]: What I meant is best shown here: 
https://github.com/avocado-framework/avocado/pull/3008/commits/485c008ee2a390d715bea1c6cbdd6c8995774fd0

File: avocado/core/safeloader.py (hunk 2 above)
Comment (ldoktor) [reply avocado/core/safeloader.py]: OK, included it in rebased version.
